### ISSUE TITLE
LivenessProbe Failure on Initial Deployment

### ISSUE THREAD

**Issue by LLParse** [kubernetes-helm/monocular#251]
The current behavior of the API pods is to fetch chart data from configured repositories (remote, by default) on first start. The API server doesn't open a socket until after this initialization occurs.

This is problematic, because net links can be really fast, or really slow. It's really hard to predict how long the initial fetch will take. Relevant band-aid PRs for users who ran into this:
https://github.com/helm/monocular/pull/239
https://github.com/helm/monocular/pull/240

Another band-aid/optimization would be to persist the cache across container restarts by defining an emptyDir volume and mounting it to `/root/monocular/repo-data`. The cache will exist for the lifetime of the pod.

The best solution for initialization of non-deterministic duration is to use init containers.
https://kubernetes.io/docs/concepts/workloads/pods/init-containers/

If we do the initial repo fetch in an init container, we can also set a sane livenessProbe initialDelaySeconds value (maybe 15s). This means not having to wait 3 minutes for the API service to accept traffic for users with fast net links or local chart repositories.

cc @galal-hussein

**Comment by prydonius** [comment]
cc @migmartri 

+1 for moving the initial repo indexing to an initContainer. There is another issue here, in that each replica is creating it's own cache. Perhaps we could move repo indexing to a scheduled job (and a Helm pre-install hook) and mount a read-only volume in each pod?

**Comment by LLParse** [comment]
Doing the repo fetching/indexing in exactly one place would work, but would introduce a single point of failure. It would also require shared storage, since pods consuming the indexed content would ideally be scheduled to different hosts for resilience.

Each replica creating its own cache is not a huge deal, in my opinion. Problems may be encountered when new charts are released; a coordination mechanism could solve that. Otherwise, it is a simple and elegant design.

**Comment by prydonius** [comment]
@LLParse good points, I agree that it's fine to have each replica create its own cache.

**Comment by vvangestel** [comment]
Could the repo sync be a background process. Currently, pulling the kubernetes stable repo takes way too long before the application becomes usable, I don't see how an initContainer would solve this.

Additionally, meta-data could be downloaded on demand when a user opens the chart info page (identifying which charts + versions are available can be found in the index.yaml, which should take no time to download). Only icons would be delayed, but I don't really mind.

**Comment by prydonius** [comment]
Just FYI, we rearchitected this in [kubeapps](https://github.com/kubeapps/kubeapps), which uses CronJobs for the syncs so that they are completed in the background. There is more context around why we started Kubeapps: Monocular was built to serve what is now https://hub.kubeapps.com and all the in-cluster features have been tacked-on. Kubeapps was our solution to an in-cluster only project, allowing us to focus on that usecase.

I expect that eventually Monocular users will migrate to kubeapps to solve their in-cluster needs, the feature we implement going forth in Monocular will be to serve the public discovery/aggregation site use-case.
